### ISSUE TITLE
Wrong behavior from no-array-index-key

### ISSUE THREAD

**Issue by MathiasKandelborg** [yannickcr/eslint-plugin-react#2727]
Hey there, I love using the plugin, it makes it so easy to write great code!

I was writing a menu when I got an error from this code:

```typescript
drawerItems.forEach((item, i) => {
    MenuComponents.push(
      <MenuItem
        key={`menu-item-${i}`}
        text={item.text[0].content}
        route={item.route}
      />
    )
    // eslint-disable-next-line no-console
  }, console.timeEnd('Menu items created'))
```

The error states that I shouldn't use index as key, even though it is a perfectly valid way of using index to make a **unique** key.

I have done this before, but after some months break time and the community shifting to use only eslint also for typescript(awesome thing btw); it seems either my settings or some defaults have changed. I can't remember how it worked before but I think the rule could be improved.

I'd love to be able to use index as a key in a React component, it could work like this: 

1. It has to be in a template string
1.1 It has to be accompanied by text.

---

The rule should probably stay as it is, but with the added possibility of using index in a template string with other text.

**Comment by ljharb** [comment]
If you're using the index, there's no point in using anything else. In other words, do `key={i}` and override the rule.

Otherwise, the proper thing to do is to not use the index *in any way whatsoever*, and to use characteristics of the item itself to derive the unique key (iow, it is *not* valid to use the index).
